bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -503,40 +501,6 @@ impl Connection {
                 }
             }
         }
-
-        loop {
-            // check if an async task exists and poll it to completion
-            if let Some(fut) = self.poll_async_task() {
-                match fut {
-                    Poll::Ready(Ok(())) => {
-                        // happy case:
-                        // continue and call s2n_negotiate to make progress on the handshake
-                    }
-                    Poll::Ready(Err(err)) => {
-                        // error case:
-                        // if the callback returned an error then abort the handshake
-                        return Poll::Ready(Err(err));
-                    }
-                    Poll::Pending => return Poll::Pending,
-                }
-            }
-
-            let res = unsafe { s2n_negotiate(self.connection.as_ptr(), &mut blocked).into_poll() };
-
-            match res {
-                Poll::Ready(res) => {
-                    let res = res.map(|_| self);
-                    return Poll::Ready(res);
-                }
-                Poll::Pending => {
-                    // if there is no connection_future then return, otherwise continue
-                    // looping and polling the future
-                    if self.context_mut().async_callback.is_none() {
-                        return Poll::Pending;
-                    }
```

Comment:
Yes. The application needs to call again on Pending.

Before, negotiate had a loop. I thought the loop made it had to read, because it wasn't clear that it only looped if blocked by an async callback. So my version for renegotiate used recursion instead of a loop. James [thought that was odd too](https://github.com/aws/s2n-tls/pull/4668#discussion_r1695640646), so I removed it. I think it's cleaner if we just let the application handle calling again.